frontend/cs/r1cs/builder.go

```diff
@@ -88,80 +114,88 @@ func newBuilder(field *big.Int, config frontend.CompileConfig) *r1cs {
 // the wire's id to the number of wires, and returns it
-func (system *r1cs) newInternalVariable() compiled.LinearExpression {
+func (system *r1cs[E, ptE]) newInternalVariable() compiled.LinearExpression[E, ptE] {
 	idx := system.NbInternalVariables + system.NbPublicVariables + system.NbSecretVariables
 	system.NbInternalVariables++
-	return compiled.LinearExpression{
-		compiled.Pack(idx, compiled.CoeffIdOne, schema.Internal),
+	return compiled.LinearExpression[E, ptE]{
+		compiled.PackInt64[E, ptE](idx, 1, schema.Internal),
 	}
 }
 
-func (system *r1cs) VariableCount(t reflect.Type) int {
+func (system *r1cs[E, ptE]) VariableCount(t reflect.Type) int {
 	return 1
 }
 
 // AddPublicVariable creates a new public Variable
-func (system *r1cs) AddPublicVariable(f *schema.Field) frontend.Variable {
+func (system *r1cs[E, ptE]) AddPublicVariable(f *schema.Field) frontend.Variable {
 	idx := len(system.Public)
 	system.Public = append(system.Public, f.FullName)
-	return compiled.LinearExpression{
-		compiled.Pack(idx, compiled.CoeffIdOne, schema.Public),
+	return compiled.LinearExpression[E, ptE]{
+		compiled.PackInt64[E, ptE](idx, 1, schema.Public),
 	}
 }
 
 // AddSecretVariable creates a new secret Variable
-func (system *r1cs) AddSecretVariable(f *schema.Field) frontend.Variable {
+func (system *r1cs[E, ptE]) AddSecretVariable(f *schema.Field) frontend.Variable {
 	idx := len(system.Secret) + system.NbPublicVariables
 	system.Secret = append(system.Secret, f.FullName)
-	return compiled.LinearExpression{
-		compiled.Pack(idx, compiled.CoeffIdOne, schema.Secret),
+	return compiled.LinearExpression[E, ptE]{
+		compiled.PackInt64[E, ptE](idx, 1, schema.Secret),
 	}
 }
 
-func (system *r1cs) one() compiled.LinearExpression {
-	return compiled.LinearExpression{
-		compiled.Pack(0, compiled.CoeffIdOne, schema.Public),
+func (system *r1cs[E, ptE]) one() compiled.LinearExpression[E, ptE] {
+	return compiled.LinearExpression[E, ptE]{
+		compiled.PackInt64[E, ptE](0, 1, schema.Public),
 	}
 }
 
 // reduces redundancy in linear expression
 // It factorizes Variable that appears multiple times with != coeff Ids
 // To ensure the determinism in the compile process, Variables are stored as public∥secret∥internal∥unset
 // for each visibility, the Variables are sorted from lowest ID to highest ID
-func (system *r1cs) reduce(l compiled.LinearExpression) compiled.LinearExpression {
+func (system *r1cs[E, ptE]) reduce(l compiled.LinearExpression[E, ptE]) compiled.LinearExpression[E, ptE] {
 	// ensure our linear expression is sorted, by visibility and by Variable ID
 	if !sort.IsSorted(l) { // may not help
 		sort.Sort(l)
 	}
+	omittable := make([]int, 0, len(l))
 
-	mod := system.Field()
-	c := new(big.Int)
 	for i := 1; i < len(l); i++ {
 		pcID, pvID, pVis := l[i-1].Unpack()
 		ccID, cvID, cVis := l[i].Unpack()
 		// if the coefficient is zero, we remove it
-		if ccID == compiled.CoeffIdZero {
-			l = append(l[:i], l[i+1:]...)
+		if l[i].IsZero() {
```

Comment:
why do we have coefficients at zero here? I can understand this to happen **after** the reduce, since we could add / sub coeffs with the result == 0, but before ?

Reply:
Not fully sure, didn't think about it. I guess this may have happened after reduce but for some reason the check was here? Makes sense to move it to post-reduce. Then would have to count nb of zeros though because now construct the new linear expression in one pass/allocation.